### DISCUSSION TITLE
2 cct light on one eps/Tasmota

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [arendst/Tasmota#18076] [Q&A]
Is it possible to configure more than one CCT (cold and warm white) light in Tasmota as CCT lights? I want to configure 4, 6, or 8 PWM channels for 2, 3, or 4 CCT lights. Tasmota determines the light type based on the number of PWM channels, or you can set SetOption68 to 1 for individual PWMs.
What option i have to set to get this behaviur
Is there a Limit of one "CCT channel"?

**Comment by barbudor** [comment]
Tasmota can handle only 1 "Light" device (ie a device that will respond to light commands such as color, dimmer, ....)

You can split the PWM as individual but that would not be "Lights" in the sense of Tasmota but you can recreate a Light management in your home automation system that would present multiple Lights user interfaces and talks PWM to Tasmota

**Comment by ChrisVonFrosta** [comment]
Too bad. But multiple relays don't represent the same device/light, right?
I'm aware that Tasmota is an MQTT gateway. However, it's now so much more than that. And even if it's just that: it makes a difference whether I have to mix 2 CCTs from 4 PWMs in the home automation system, or if I can set the color twice on one ESP. Otherwise, I'd need an ESP for each CCT channel, in which case I'd need 4+x ESPs per floor, instead of one ESP centrally located on an existing board (16 channel relay, 16 channel input, ...). This can be unnecessary for multiple reasons, if you know what I mean.

If Sonoff were to release a 2-channel CCT LED driver, would I have a better chance of implementation? Just kidding ;)
Can I enhance Tasmota to a fictional Sonoff CCT multi-channel driver with a reasonable effort?
Thank you!

**Reply by sfromis** [reply]
You could do something "CCT-like" by configuring 5 PWM channels (not 4), where the last two are what Tasmota recognizes as CCT, and then use two of the first 3 for the other light, where varying the color balance would vary the color temperature, as long as you have split CCT from RGB. Not ideal, but likely a bit easier....

With ESP32, you can have additional PWM channels outside of the 5 used for the Tasmota light feature, and control them with a bit of Berry code for CCT.

**Reply by ChrisVonFrosta** [reply]
Thank you so much that you thinking along! But i think that doesn't match my search.
Both of yours suggestions  need to "waste" gpios if i can't use them as PWM?
also it would be no nicly structure and labelling.
i just started with berry and i would be fine with using it. But dont like the Ratio 5 for 2 PWM Channel (even gpios).

i have led strips which include both, warm and cold white led. each white with seprat negaive trace
+24V ------
- ww ------- 
- cw  ------- 

i want to connect 2 or more of those as seperate lights on one ESP (with mosfet driver in between)
It would be greate to have the brightnesss and whiteness sliders + button for each of them in the webgui
like separate PWM channels, but in pairs of two (for cct).

i never dived to deep in the source code.. i did some custom config builds but never edited the programm it self.
how diffcult would it be to implement multi (cct) pwm lights by my self? where is the point i should locking at?
Or is  the any other option?

**Reply by sfromis** [reply]
In my understanding, the people knowing how the lights feature work shy away from messing with it, due to it being intricate to make combinations work. Tasmota just does not support multiple lights (with full Tasmota light features for each), and not easy to "just add".

A couple of ways forward:
- Lots of work to make it work the hard way, amount depends on how "nice" you want it
- Less messy with ESP32 and Berry, but still a non-trivial amount of work to code it. Berry code only has full control over PWM6-8, (or less with ESP32-C3)
- Get more than one ESP. Easy to do, and does not cost much.

I consider it a waste of time to worry a lot about "wasting" gpios, when you do not have "lots" of PWM channels potentially available anyway - unless it is only for the heck of it :stuck_out_tongue_winking_eye:

**Comment by ChrisVonFrosta** [comment]
i have this [board](https://de.aliexpress.com/item/1005004292634218.html?spm=a2g0o.order_list.order_list_main.27.38525c5fjZRDw0&gatewayAdapt=glo2deu)
I want to control all ceiling lights at home with and connect all corresponding wall switches to it. But in a way that a standard function is still guaranteed even without a network or home automation center. With that board and Tasmota, I can implement this completely and centrally with one board and one Tasmota ESP. As long as I stick with relays, which is allready great, I cannot dim the 24V CCT strips or influence their color. I'm hesitant to outsource this to a multitude of external ESPs because of the many unnecessary Wi-Fi clients and configurations.

I am actually willing to take a look at whether I can adapt or expand the firmware accordingly. I know in functional C (Atmega/Linux) -  more or less. Less in objectiv C. However, I have not been able to grasp this extensive C++ program. I have some knowledge in event-driven programming from JavaScript. If you could show me a good starting point, I would look into it to see if I can realistically make a contribution. If so, I would of course make it available here if desired.
Which files are relevant?

My layman's approach would be to map the PWM outputs differently with a new Setoption, specifically in pairs of two (brightness and color/temp). If PWM is implemented only as a single light or as separate PWM channels, the most difficult work would likely be expected here, right? Perhaps you could easily estimate whether it would be easier to expand the separate channels or increase the number of "one" lamps?

Maybe this is a good reason to dive into your project.

Perhaps I would prefer BerryScript - at least I could get it implemented - but if you're saying that only 3 channels are available with it, then that doesn't help me much.

**Reply by sfromis** [reply]
Why would you even see a relay board as a relevant starting point for lots of PWM channels? What hardware features would it have to drive PWM channels. I see nothing.

**Comment by ChrisVonFrosta** [comment]
That board was the starting point. The idea of PWM for lights was born later, when i saw how good this works. i never thought about before that PWM-lights are limited to one.
I want to add a PWM extension via I2C to this board, or perhaps, I would buy/create a version with a mix of MOSFET outputs and relays. The relays control all non-dimmable lights or some special power sockets.
The PWM should be for the 24V LEDs, so I could have multiple PWMs in addition to the relays. It would be fantastic! The Board also have inputs (for the wall switches), so it would be the complete all-one- solution. I can't see the difference in paradigms between relays and PWMs when it comes to the number of possible lights.

I have now activated 16 relays. However, that doesn't correspond to "one" light. It's not a problem with relays or outputs. You can also connect 20 temperature sensors. Therefore, I don't understand the aversion to multiple PWM lights on a TasmotaEsp. Does this aversion stem from it not having been implemented yet, or is it not desirable under any circumstances in Tasmota? Maybe I'm just too stupid to get it ;)

**Reply by ChrisVonFrosta** [reply]
Just because of the inputs with optocouplers where I connect the wall switches and because of the Ethernet connection, I want to stick with the ONE ESP variant. It would be just perfect with that one feature. Please excuse me if I am repeating myself.

A feature request probably doesn't make much sense, does it?
I would really like to try to implement it. It will be even harder for me without tips ;)

**Reply by ChrisVonFrosta** [reply]
I can envision a solution with Berry and individual PWM outputs, provided that I can control all the necessary PWM channels. However, I would also need to implement the color mixing myself and, most importantly, I would not have the great color and brightness sliders per light in the web GUI. Additionally, addressing it via MQTT from a home automation system would be less convenient, since I would not have a topic for the colors or brightness of the lamps, but only "anonymous" PWM channels. I cannot estimate the additional CPU load, but it would certainly be worse, wouldn't it?

**Reply by sfromis** [reply]
As ESP32 boards are dirt cheap, I do not see is as a good plan to try build things around a specialized board using ESP32, unless you do it using the PCA9685 recommended by barbudor, and manage CCT mixing of channels, web UI, and so on yourself.

**Comment by ChrisVonFrosta** [comment]
That is not a question of price. I also do not see it as a board-specific solution... being able to configure multiple CCT lights is just as generic as multiple relays, isn't it?
Where is the philosophical difference between relays and PWM outputs in this context?

You don't see the disadvantages of having multiple WiFi ESPs compared to this one board connected via Ethernet?

- chance/initial configuration 1 vs x clients
- 1 Ethernet vs x WiFi clients
- control all lights off: 1 command vs x commands on x clients
- integration into smart home: 1 vs x clients
- hardware expansion: 1 vs x
    

Multiple ESPs bring many disadvantages. I have an ESP board connected via Ethernet, centrally located in the fuse box, instead of multiple WiFi ESPs without the necessary peripherals (power supply, optocouplers for inputs, MOSFETs for outputs, etc.). So I would have to buy and solder peripherals for each ESP... and then it would be a poorly maintainable system compared to such a nice central board. Since all the lines (lighting, wall switches) lead to the central box anyway, I don't want to install 10+x ESPs there. The board is on a DIN rail and everything would be neatly wired.

However, multiple ESPs are not an option. At the very beginning, the idea was to distribute Sonoff minis in the rooms. But that is not possible because of the existing wiring, and for other reasons, it is not desired.

I would prefer to use the PCA9685 connected via i2c.  When you say manage by myself, you mean with berry script, don't you?

**Reply by barbudor** [reply]
Chris
I can understand your point of having everything centralized while myself is more in favor of distributed

Pca9685 and Berry soundd the best for me because your can build a totaly new workflow, your own commands (like XDimmer instead of Dimmer meaning eXtended Dimmer)
Trying to extend tasmota to support more PWM and more light is probably possible but
- not easy
- probably impossible to do it without breaking the current Settings structure
- which means it wouldn't be merged in the trunk and you will end up with a fork that would drift from the core.

With Berry you are pretty much independant although may be more to re-do but probably easier than trying to inject your changes in Tasmota.

What about the hardware?
Have you already identified a board that would offer 16PWM native or PCA? Or are you planning to design your own?

**Reply by sfromis** [reply]
> I would prefer to use the PCA9685 connected via i2c. When you say manage by myself, you mean with berry script, don't you?

Yeah, I specifically voiced my agreement with using that chip. "Manage by yourself" is by whatever means you choose as feasible, including coding in Berry, but I deliberately did not focus on one specific option. However, Berry is probably the most approachable way, unless you aim for "going deeper" into creating a native Tasmota driver, or wrangle with how complicated it may get to significantly change the Tasmota lights driver for a multi-pronged setup.

In any case, if you want it easy, multiple ESP chips is easiest, could be on one single board, and to add a bunch of MOSFET you'd probably want to design a PCB for that, anyway. You'd probably also need that if using PCA9685.

**Comment by ChrisVonFrosta** [comment]
I have no clue how much workload it could be to add a driver for multiple lights natively. I think it's even more complicated if I go with the extension via PCA9685. It sounds like two changes are needed: the light structure (to allow multiple lights) and the addition of PCA9685 to multiple lights.

I have the board that I linked here at home. Since the PWM idea came later, I don't have another board here yet. But the PWM extension module and additional Mosfet board are on the way. First, I want to make sure that I can realize my concept with a reasonable amount of effort.  So, for now, I will extend the existing board with plug-in extensions. If it becomes clear that it works, then I will design my own board if the manufacturer (or another) does not offer a board variant with Mosfets via PCA9685.

**Comment by ChrisVonFrosta** [comment]
How to check the pushed button? is ther something like interrupt or doe i have to poll with gpio.read()?
can i create sliders in webgui to adjust light properties (variables in berry script)? 

In Berry i think this way (just a draft):
what do you think?
```
class cct_light
    var id               # id of the cct light
    var name             # name of the cct light
    var port_ww, port_cw # ports for warm white and cool white lights respectively
    var brightnes, whiteness # brightness and whiteness of the light
    
     def init(id,name, port_ww, port_cw) # constructor for the cct light class
        self.id=id
        self.name=name
        self.port_ww=port_ww
        self.port_cw=port_cw
    end


    def set_power(value) # method to set the power of the cct light
        ...
    end

    def toggle_power() # method to toggle the power of the cct light
        ...
    end

    def set_brightness(value) # method to set the brightness of the cct light
        ...
    end
    
    def set_whiteness(value) # method to set the whiteness of the cct light
        ...
    end

    def proceed_button_push(type) # method to proceed with button push action

        if type == "single_psuh_short" 
            self.toggle_power()
        end
    end
    
end


class button
    var id             # id of the button
    var portpin        # pin of the button
    var push_type      # type of button push (single push, long push etc.)
    var related_relays = [] # list of related relays
    
    def init(id, portpin,[]) # constructor for the button class
        self.id = id
        self.portpin = portpin
        self.related_relays = related_relays
    end
    
    
    def is_pressed() # method to check if the button is pressed
        for i: 0...20
            if gpio.digital_read(self.portpin)
                self.push_type="single_psuh_short"
                return true
            end
            tasmota.delay(10ms)
        end
    end

end



def which_button_pressed(buttons) # function to determine which button is pressed
    var relay_id=nil
    var push_type=nil
    
    for i: 0 ... list.size(buttons)     # check all buttons

        if buttons[i].is_pressed()      # this i button ist pressed
            
            var size = button[i].related_relays.size()  # how many relay are related to this button
            for j: 0..size
                relay_id = button[i].related_relays[j]
                push_type =button[i].type
                cct_lights[relay_id].proceed_button_push(push_type)
            end
        end
    end
end


# list of cct lights with their id, name, and port information
cc_lights = [cct_light(1, "kitchen", 10,11), cct_light(2, "floor", 12,13), cct_light(3, "bathroom", 14,15), cct_light(4, "other", 16,17)]

# list of buttons with their id, port, and related relays
buttons = [button(1,20,[1], button(2,21,[2], button(3,22,[2,4]] 


def tasmota.interrupt.pinchage()) # interrupt function
    on(....)
       which_button_pressed(buttons) # determine which button is pressed
    end
end
```

**Comment by sfromis** [comment]
> How to check the pushed button?

The usual `tasmota.add_rule(....)`

> can i create sliders in webgui

Others have created sliders by coding the HTML/JS structures needed to make that work. I've not dabbled in that.

**Comment by beobuild** [comment]
I'm having the same idea, any other new workarounds?

**Reply by sfromis** [reply]
Nothing changed about Tasmota only supporting one CCT light.
